packages/contracts-bedrock/src/L2/SuperchainERC20Bridge.sol

```diff
@@ -0,0 +1,99 @@
+    error InvalidCrossDomainSender();
+
+    /// @notice Emitted when tokens are sent from one chain to another.
+    /// @param token         Address of the token sent.
+    /// @param from          Address of the sender.
+    /// @param to            Address of the recipient.
+    /// @param amount        Number of tokens sent.
+    /// @param destination   Chain ID of the destination chain.
+    event SendERC20(
+        address indexed token, address indexed from, address indexed to, uint256 amount, uint256 destination
+    );
+
+    /// @notice Emitted whenever tokens are successfully relayed on this chain.
+    /// @param token         Address of the token relayed.
+    /// @param from          Address of the msg.sender of sendERC20 on the source chain.
+    /// @param to            Address of the recipient.
+    /// @param amount        Amount of tokens relayed.
+    /// @param source        Chain ID of the source chain.
+    event RelayERC20(address indexed token, address indexed from, address indexed to, uint256 amount, uint256 source);
+
+    /// @notice Address of the L2ToL2CrossDomainMessenger Predeploy.
+    address internal constant MESSENGER = Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER;
+
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0-beta.1
+    string public constant version = "1.0.0-beta.1";
+
+    /// @notice Sends tokens to a target address on another chain.
+    /// @dev Tokens are burned on the source chain.
+    /// @param _token    Token to send.
+    /// @param _to       Address to send tokens to.
+    /// @param _amount   Amount of tokens to send.
+    /// @param _chainId  Chain ID of the destination chain.
+    /// @return msgHash_ Hash of the message sent.
+    function sendERC20(
+        address _token,
+        address _to,
+        uint256 _amount,
+        uint256 _chainId
+    )
+        external
+        returns (bytes32 msgHash_)
+    {
+        if (_to == address(0)) revert ZeroAddress();
+
+        ISuperchainERC20(_token).__crosschainBurn(msg.sender, _amount);
+
+        bytes memory message = abi.encodeCall(this.relayERC20, (_token, msg.sender, _to, _amount));
+        msgHash_ = IL2ToL2CrossDomainMessenger(MESSENGER).sendMessage(_chainId, address(this), message);
+
+        emit SendERC20(_token, msg.sender, _to, _amount, _chainId);
+    }
+
+    /// @notice Relays tokens received from another chain.
+    /// @dev Tokens are minted on the destination chain.
+    /// @param _token   Token to relay.
+    /// @param _from    Address of the msg.sender of sendERC20 on the source chain.
+    /// @param _to      Address to relay tokens to.
+    /// @param _amount  Amount of tokens to relay.
+    function relayERC20(address _token, address _from, address _to, uint256 _amount) external {
+        if (msg.sender != MESSENGER) revert CallerNotL2ToL2CrossDomainMessenger();
+
+        if (IL2ToL2CrossDomainMessenger(MESSENGER).crossDomainMessageSender() != address(this)) {
```

Comment:
Would you find any value in adding a `context()` function that returns both the source + sender + other contextual information? I notice we do this double call thing everywhere, it could be a simplification

Reply:
It is a useful change, but we will be doing it in another PR to reduce the changes scope